content/advanced/340_appmesh_flagger/canary_analysis.md

```diff
@@ -0,0 +1,195 @@
+
+#### Canary Resource
+
+[Flagger](https://docs.flagger.app/) can be configured to automate the release process for Kubernetes workloads with a custom resource named canary which we installed in the previous chapter. 
+The canary custom resource defines the release process of an application running on Kubernetes. 
+
+We have defined the canary release with progressive traffic shifting for the deployment of backend service `detail` [here](https://github.com/aws-containers/eks-app-mesh-flagger-demo/blob/main/deployment/flagger-canary.yaml).
+
+When we deploy a new version of `detail` backend service, Flagger gradually shifts traffic to the canary, and at the same time, measures the requests success rate as well as the average response duration.
+
+#### Canary Target
+
+Below is one of the section from Canary Resource for canary target for `detail` service. For more details, see Flagger documentation [here](https://docs.flagger.app/usage/how-it-works#canary-target).
+
+{{< output >}}  
+spec:
+  targetRef:
+    apiVersion: apps/v1
+    kind: Deployment
+    name: detail
+  progressDeadlineSeconds: 60
+{{< /output >}}
+
+Based on the above configuration, Flagger generates the following Kubernetes object:
+
+- deployment/detail-primary
+
+The primary deployment is considered the stable release of our `detail` service, by default all traffic is routed to this version and the target deployment is scaled to zero. Flagger will detect changes to the target deployment (including secrets and configmaps) and will perform a canary analysis before promoting the new version as primary.
+
+The progress deadline represents the maximum time in seconds for the canary deployment to make progress before it is rolled back, defaults to ten minutes.
+
+#### Canary Service
+
+Below canary service section from Canary Resource dictates how the target workload is exposed inside the cluster. The canary target should expose a TCP port that will be used by Flagger to create the ClusterIP Services. For more details, see Flagger documentation [here](https://docs.flagger.app/usage/how-it-works#canary-service).
+
+{{< output >}}  
+  service:
+    # container port
+    port: 3000
+{{< /output >}}
+
+Based on our [canary spec](https://github.com/aws-containers/eks-app-mesh-flagger-demo/blob/main/deployment/flagger-canary.yaml) service, Flagger creates the following Kubernetes ClusterIP service
+* `detail.flagger.svc.cluster.local`
+  
+  selector `app=detail-primary`
+* `detail-primary.flagger.svc.cluster.local`
+
+  selector `app=detail-primary`
+* `detail-canary.flagger.svc.cluster.local`
+
+  selector `app=detail`
+
+This ensures that traffic to `detail.flagger:3000` will be routed to the latest stable release of our `detail` service. The `detail-canary.flagger:3000` address is available only during the canary analysis and can be used for conformance testing or load testing.
+
+
+#### Canary Analysis
+
+The canary analysis defines:
```

Comment:
I agree but we still need to give an abstract idea of canary analysis is rather than making user to refer the original document where everything may not be relevant. And I have also referenced the original flagger documentation each and every place in the entire workshop wherever it's necessary.